sdk/resourcemanager/azure-resourcemanager-appplatform/src/main/java/com/azure/resourcemanager/appplatform/implementation/SpringAppDeploymentImpl.java

```diff
@@ -590,54 +586,33 @@ public Mono<Indexable> apply(Context context) {
-                    ResourceUtils.nameFromResourceId(context.getData("buildId")));
-        }
-
-        private Mono<PollResponse<PollResult<BuildInner>>> waitForBuild(PollingContext<PollResult<BuildInner>> context) {
-            return getBuildResult(context)
-                .flatMap((Function<BuildResultInner, Mono<PollResponse<PollResult<BuildInner>>>>) buildResultInner -> {
-                    BuildResultProvisioningState state = buildResultInner.properties().provisioningState();
-                    // construct empty result, "buildId" from context is the real result
-                    PollResult<BuildInner> emptyResult = new PollResult<>(new BuildInner().withProperties(new BuildProperties()));
-                    if (state == BuildResultProvisioningState.SUCCEEDED) {
-                        return Mono.just(new PollResponse<>(LongRunningOperationStatus.SUCCESSFULLY_COMPLETED, emptyResult));
-                    } else if (state == BuildResultProvisioningState.FAILED || state == BuildResultProvisioningState.DELETING) {
-                        return Mono.error(new RuntimeException("build failed"));
-                    } else if (state == BuildResultProvisioningState.QUEUING) {
-                        return Mono.just(new PollResponse<>(LongRunningOperationStatus.NOT_STARTED, emptyResult));
-                    }
-                    return Mono.just(new PollResponse<>(LongRunningOperationStatus.IN_PROGRESS, emptyResult));
-                });
-        }
-
-        private Mono<PollResult<BuildInner>> enqueueBuild(ResourceUploadDefinition option, PollingContext<PollResult<BuildInner>> context) {
+        private Mono<String> uploadAndBuildAsync(File source, ResourceUploadDefinition option) {
+            return uploadToStorageAsync(source, option)
+                .then(enqueueBuildAsync(option))
+                .flatMap(buildId ->
+                    manager().serviceClient().getBuildServices()
+                        .getBuildResultAsync(
+                            service().resourceGroupName(),
+                            service().name(),
+                            Constants.DEFAULT_TANZU_COMPONENT_NAME,
+                            parent().name(),
+                            ResourceUtils.nameFromResourceId(buildId))
+                        .flatMap(buildResultInner -> {
+                            BuildResultProvisioningState state = buildResultInner.properties().provisioningState();
+                            if (state == BuildResultProvisioningState.SUCCEEDED) {
+                                return Mono.just(buildId);
+                            } else if (state == BuildResultProvisioningState.QUEUING || state == BuildResultProvisioningState.BUILDING) {
+                                return Mono.empty();
+                            } else return Mono.error(new RuntimeException("build failed"));
+                        }).repeatWhenEmpty(
```

Comment:
Do we need a timeout? A hour is also fine but I guess we need one?

Can be added later.

Reply:
Sure, I'll add timeout.